### PR TITLE
[feature][array-type]support select ARRAY data type on vectorized engine

### PR DESCRIPTION
**Pull request by cambyzju**
# Proposed changes

Issue Number: close #7570

## Problem Summary:

Usage Example:
1. create table for test;
`CREATE TABLE `array_test` (
  `k1` tinyint(4) NOT NULL COMMENT "",
  `k2` smallint(6) NULL COMMENT "",
  `k3` ARRAY<int(11)> NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`k1`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k1`) BUCKETS 5
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"in_memory" = "false",
"storage_format" = "V2"
);`

2. insert some data
`insert into array_test values(1, 2, [1, 2]);`
`insert into array_test values(2, 3, null);`
`insert into array_test values(3, null, null);`
`insert into array_test values(4, null, []);`

3. open vectorized
`set enable_vectorized_engine=true;`

4. query array data
`select * from array_test;`
+------+------+--------+
| k1   | k2   | k3     |
+------+------+--------+
|    4 | NULL | []     |
|    2 |    3 | NULL   |
|    1 |    2 | [1, 2] |
|    3 | NULL | NULL   |
+------+------+--------+
4 rows in set (0.061 sec)

Code Changes include：
1. add column_array, data_type_array codes;
2. codes about data_type creation by Field, TabletColumn, TypeDescriptor, PColumnMeta move to DataTypeFactory;
3. support create data_type for ARRAY date type;
4. RowBlockV2::convert_to_vec_block support ARRAY date type;
5. VMysqlResultWriter::append_block support ARRAY date type;
6. vectorized::Block serialize and deserialize support ARRAY date type;

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morningman** [comment]
Can we do the select array after this PR is merged?
Could you give some sql example?

**Comment by cambyzju** [comment]
> Can we do the select array after this PR is merged? Could you give some sql example?

yes.
sql example have added in this pr.

File: be/src/olap/row_block2.cpp [hunk 1]
```diff
@@ -328,6 +348,222 @@ Status RowBlockV2::_copy_data_to_column(int cid, doris::vectorized::MutableColum
...
+                    value.deserialize(slice->data);
+                    *pvalue = std::move(value);
+                } else {
+                    *pvalue = std::move(*reinterpret_cast<BitmapValue*>(slice->data));
+                }
+            }
+        }
+        break;
+    }
+    case OLAP_FIELD_TYPE_HLL:
```
Comment (zhangstar333) [inline be/src/olap/row_block2.cpp:413]: now in vectorized mode, hll type have done refactor, could look in this pr #7828

File: be/src/vec/columns/column_array.cpp [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+
+void ColumnArray::insert_default() {
+    /// NOTE 1: We can use back() even if the array is empty (due to zero -1th element in PODArray).
+    /// NOTE 2: We cannot use reference in push_back, because reference get invalidated if array is reallocated.
+    auto last_offset = get_offsets().back();
+    get_offsets().push_back(last_offset);
+}
+
+void ColumnArray::pop_back(size_t n) {
+    auto & offsets_data = get_offsets();
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/vec/columns/column_array.cpp:259]: done

File: be/src/vec/columns/column_array.cpp [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+    offsets_data.resize_assume_reserved(offsets_data.size() - n);
+}
+
+void ColumnArray::reserve(size_t n) {
+    get_offsets().reserve(n);
+    get_data().reserve(n); /// The average size of arrays is not taken into account here. Or it is considered to be no more than 1.
+}
+
+size_t ColumnArray::byte_size() const {
+    return get_data().byte_size() + get_offsets().size() * sizeof(get_offsets()[0]);
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/vec/columns/column_array.cpp:273]: get_offsets.byte_size() = (get_offsets().size() **+ 1**) * sizeof(get_offsets()[0]);

File: be/src/olap/row_block2.cpp (hunk 1 above)
Comment (cambyzju) [reply be/src/olap/row_block2.cpp]: done

File: be/src/vec/columns/column_array.cpp [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+        size_t prev_max_offset = old_size ? cur_offsets.back() : 0;
+        cur_offsets.resize(old_size + length);
+
+        for (size_t i = 0; i < length; ++i)
+            cur_offsets[old_size + i] = src_offsets[start + i] - nested_offset + prev_max_offset;
+    }
+}
+
+ColumnPtr ColumnArray::filter(const Filter & filt, ssize_t result_size_hint) const {
+    if (typeid_cast<const ColumnUInt8 *>(data.get()))      return filter_number<UInt8>(filt, result_size_hint);
```
Comment (HappenLee) [inline be/src/vec/columns/column_array.cpp:323]: Support `ColumnComplex `

File: be/src/vec/columns/column_array.cpp [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+
+    return ColumnArray::create(
+        ColumnNullable::create(
+            filtered_array_of_nested.get_data_ptr(),
+            std::move(res_null_map)),
+        filtered_offsets);
+}
+
+void ColumnArray::insert_indices_from(const IColumn& src, const int* indices_begin, const int* indices_end) {
+    for (auto x = indices_begin; x != indices_end; ++x) {
```
Comment (HappenLee) [inline be/src/vec/columns/column_array.cpp:480]: consider `index == -1`

File: be/src/vec/columns/column_array.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/vec/columns/column_array.cpp:273]: PodArray only has a static function byte_size(num_of_elements).

File: be/src/vec/columns/column_array.cpp (hunk 4 above)
Comment (cambyzju) [reply be/src/vec/columns/column_array.cpp:323]: next pr will support.

File: be/src/vec/columns/column_array.cpp (hunk 5 above)
Comment (cambyzju) [reply be/src/vec/columns/column_array.cpp:480]: done

**Review by morningman — APPROVED** [review]
LGTM

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 6]
```diff
@@ -723,14 +723,8 @@ void SegmentIterator::_init_current_block(
         for (size_t i = 0; i < _schema.num_column_ids(); i++) {
             auto cid = _schema.column_id(i);
             auto column_desc = _schema.column(cid);
-            auto data_type = Schema::get_data_type_ptr(column_desc->type());
-            if (column_desc->is_nullable()) {
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:727]: Schema::get_data_type_ptr will consider is_nullable, and will create DataTypeNullable if needed.

**Pull request merged**